controllers/cloud.redhat.com/clowdapp_reconciliation.go

```diff
@@ -0,0 +1,381 @@
+)
+
+type ClowdAppReconciliation struct {
+	cache                 *rc.ObjectCache
+	recorder              record.EventRecorder
+	ctx                   *context.Context
+	client                client.Client
+	log                   *logr.Logger
+	app                   *crd.ClowdApp
+	req                   *ctrl.Request
+	env                   *crd.ClowdEnvironment
+	reconciliationMetrics ReconciliationMetrics
+}
+
+func MakeClowdAppReconciliation(ctx context.Context,
+	cache *rc.ObjectCache,
+	client client.Client,
+	recorder record.EventRecorder,
+	app *crd.ClowdApp,
+	log logr.Logger,
+	req *ctrl.Request) ClowdAppReconciliation {
+	return ClowdAppReconciliation{
+		cache:    cache,
+		recorder: recorder,
+		ctx:      &ctx,
+		client:   client,
+		log:      &log,
+		app:      app,
+		req:      req,
+	}
+}
+
+func (r *ClowdAppReconciliation) steps() []func() (ctrl.Result, error) {
+	return []func() (ctrl.Result, error){
+		r.getApp,
+		r.setPresentAndManagedApps,
+		r.startMetrics,
+		r.isAppMarkedForDeletion,
+		r.addFinalizer,
+		r.isEnvLocked,
+		r.isAppDisabled,
+		r.getClowdEnv,
+		r.isEnvCoreNamespaceDeleted,
+		r.isAppCoreNamespaceDeleted,
+		r.isClowdEnvReconciled,
+		r.isEnvReady,
+		r.runProviders,
+		r.applyCache,
+		r.setAppResourceStatus,
+		r.deletedUnusedResources,
+		r.setReconciliationSuccessful,
+		r.stopMetrics,
+	}
+}
+
+func (r *ClowdAppReconciliation) Reconcile() (ctrl.Result, error) {
+	for _, step := range r.steps() {
+		result, err := step()
+		if err != nil {
+			return result, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) startMetrics() (ctrl.Result, error) {
+	r.reconciliationMetrics = ReconciliationMetrics{}
+	r.reconciliationMetrics.init(r.app.Name, r.app.Spec.EnvName)
+	r.reconciliationMetrics.start()
+
+	if clowderconfig.LoadedConfig.Features.PerProviderMetrics {
+		requestMetrics.With(prometheus.Labels{"type": "app", "name": r.app.GetIdent()}).Inc()
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) stopMetrics() (ctrl.Result, error) {
+	r.reconciliationMetrics.stop()
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) setPresentAndManagedApps() (ctrl.Result, error) {
+	if _, ok := presentApps[r.app.GetIdent()]; !ok {
+		presentApps[r.app.GetIdent()] = true
+	}
+	presentAppsMetric.Set(float64(len(presentApps)))
+
+	delete(managedApps, r.app.GetIdent())
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) getApp() (ctrl.Result, error) {
+	if getAppErr := r.client.Get(*r.ctx, r.req.NamespacedName, r.app); getAppErr != nil {
+		if k8serr.IsNotFound(getAppErr) {
+			// Must have been deleted
+			return ctrl.Result{}, errors.New(SKIPRECONCILE)
+		}
+		r.log.Info("App not found", "env", r.app.Spec.EnvName, "app", r.app.GetIdent(), "err", getAppErr)
+		return ctrl.Result{}, getAppErr
+	}
+	//This is kinda side-effecty but I couldn't think of a better place
+	//to put it.
+	logWithEnv := r.log.WithValues("env", r.app.Spec.EnvName)
```

Comment:
This is a side effect of the steps. Piece, one could argue it could become its own step. It also doesn't matter when it happens, so we could move this little piece out and into the actual reconciliation function as it references the EnvName from the spec and not from a real ClowdEnv, which is only gotten half way through the steps.

Reply:
Yeah this is a tough little one. Where I wanted to put this was at the very top of `ClowdAppReconciler.Reconcile` where we do all of those `r.Log.WithValues` calls, but I couldn't figure out how to get access to the env name before getting the app.  If there is a way to get the env name before getting the app then that would be a good change I think, but I don't know how.

So then the most strictly idiomatic thing would be to put it in a step, right? Yeah, it would be, except that this log update has no conditionality associated with it. It can't (so far as I can tell) fail. Additionally, there's no implementation to test, as this is a single public method on the logging library and thus outside of scope for testing. So, in this particular case a step would buy us nothing: no improvement to testing and  no reduction in complexity.

So long story short (too late) if you can tell me a way to grab env name before loading the app I'll move it out of the step. If not then I suggest we leave it as is.